File: api/images/edit.ts
import type { VercelRequest, VercelResponse } from "@vercel/node";

const XAI_API_URL = "https://api.x.ai/v1/images/edits";

export default async function handler(req: VercelRequest, res: VercelResponse) {
  res.setHeader("Access-Control-Allow-Origin", "*");
  res.setHeader("Access-Control-Allow-Methods", "POST, OPTIONS");
  res.setHeader("Access-Control-Allow-Headers", "Content-Type");

  if (req.method === "OPTIONS") {
    return res.status(200).end();
  }
  if (req.method !== "POST") {
    return res.status(405).json({ error: "Method not allowed" });
  }

  const apiKey = process.env.XAI_API_KEY;
  if (!apiKey) {
    return res.status(500).json({ error: "XAI_API_KEY not configured" });
  }

  const { prompt, image_url } = req.body;
  if (!prompt) {
    return res.status(400).json({ error: "Missing prompt" });
  }
  if (!image_url) {
    return res.status(400).json({ error: "Missing image_url" });
  }

  try {
    const body: Record<string, unknown> = {
      model: "grok-imagine-image",
      prompt,
      image: {
        url: image_url,
        type: "image_url",
      },
    };

    const resp = await fetch(XAI_API_URL, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify(body),
    });

    if (!resp.ok) {
      const errText = await resp.text().catch(() => "");
      return res.status(resp.status).json({
        error: `xAI API error (${resp.status})`,
        details: errText.slice(0, 500),
      });
    }

    const data = await resp.json();
    return res.status(200).json(data);
  } catch (err) {
    return res
      .status(500)
      .json({ error: "Failed to call xAI API", details: String(err) });
  }
}

File: api/images/edit.py
from http.server import BaseHTTPRequestHandler
import json
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..'))
from grok_lib.client import Grok


class handler(BaseHTTPRequestHandler):

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def do_POST(self):
        try:
            content_length = int(self.headers.get('Content-Length', 0))
            body = json.loads(self.rfile.read(content_length)) if content_length else {}

            prompt = body.get('prompt')
            if not prompt:
                self._send_json(400, {"error": "Missing prompt"})
                return

            file_attachments = body.get('fileAttachments', [])

            grok = Grok(model="grok-3")
            result = grok.start_convo(
                message=f"Edit an image based on this description: {prompt}",
                enableImageGeneration=True,
                imageGenerationCount=2,
                fileAttachments=file_attachments,
                forceConcise=True,
                disableTextFollowUps=True,
            )

            if result.get('error'):
                self._send_json(500, {"error": result['error']})
                return

            image_urls = result.get("images") or []
            if image_urls:
                self._send_json(200, {
                    "data": [{"url": url} for url in image_urls],
                })
            else:
                self._send_json(200, {
                    "data": [],
                    "message": result.get("response") or "No images were generated. Try a different prompt.",
                })

        except Exception as e:
            self._send_json(500, {"error": "Failed to edit image", "details": str(e)})

    def _send_json(self, status_code, data):
        self.send_response(status_code)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(json.dumps(data).encode())
